src/base/RandomUuid.cc

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "base/IoManip.h"
+#include "base/RandomUuid.h"
+#include "base/TextException.h"
+#include "defines.h"
+
+#include <iostream>
+#include <random>
+
+static_assert(sizeof(RandomUuid) == 128/8, "RandomUuid has RFC 4122-prescribed 128-bit size");
+
+RandomUuid::RandomUuid()
+{
+    // Generate random bits for populating our UUID.
+    // STL implementation bugs notwithstanding (e.g., MinGW bug #338), this is
+    // our best chance of getting a non-deterministic seed value for the r.n.g.
+    static std::mt19937_64 rng(std::random_device{}()); // produces 64-bit sized values
+    const auto rnd1 = rng();
+    const auto rnd2 = rng();
+
+    // bullet 3 of RFC 4122 Section 4.4 algorithm but setting _all_ bits (KISS)
+    static_assert(sizeof(rnd1) + sizeof(rnd2) == sizeof(*this), "random bits fill a UUID");
+    memcpy(raw(), &rnd1, sizeof(rnd1));
+    memcpy(raw() + sizeof(rnd1), &rnd2, sizeof(rnd2));
+
+    // bullet 1 of RFC 4122 Section 4.4 algorithm
+    EBIT_CLR(clockSeqHiAndReserved, 6);
+    EBIT_SET(clockSeqHiAndReserved, 7);
+
+    // bullet 2 of RFC 4122 Section 4.4 algorithm
+    EBIT_CLR(timeHiAndVersion, 13);
+    EBIT_SET(timeHiAndVersion, 14);
+    EBIT_CLR(timeHiAndVersion, 15);
+    EBIT_CLR(timeHiAndVersion, 16);
+}
+
+RandomUuid::RandomUuid(const Serialized &bytes)
+{
+    static_assert(sizeof(*this) == sizeof(Serialized), "RandomUuid is deserialized with 128/8 bytes");
+    memcpy(raw(), bytes.data(), sizeof(*this));
+    timeHiAndVersion = ntohs(timeHiAndVersion);
+    if (!check())
+        throw TextException("malformed version 4 variant 1 UUID", Here());
+}
+
+bool
+RandomUuid::check() const
+{
+    return (!EBIT_TEST(clockSeqHiAndReserved, 6) &&
+            EBIT_TEST(clockSeqHiAndReserved, 7) &&
+            !EBIT_TEST(timeHiAndVersion, 13) &&
+            EBIT_TEST(timeHiAndVersion, 14) &&
+            !EBIT_TEST(timeHiAndVersion, 15) &&
+            !EBIT_TEST(timeHiAndVersion, 16));
+}
+
+RandomUuid::Serialized
+RandomUuid::serialize() const
+{
+    auto serialized = serialize();
+    auto *t = reinterpret_cast<uint16_t *>(&serialized[0] + offsetof(RandomUuid, timeHiAndVersion));
+    *t = htons(*t);
+    return serialized;
+}
+
+void
+RandomUuid::print(std::ostream &os) const
+{
+    PrintHex(os << "UUID:", raw(), sizeof(*this));
```

Comment:
> UUID value always uses the network byte order

A better way to put that would be "so that UUID internal structure and the corresponding implementation stop being dependent on the byte order". The RFC algorithm is byte-order dependent because it would be very difficult for them to sketch a different implementation for other UUID versions -- the versions that actually use those multibyte fields to store multibyte values like time and counters. In our subcase, version 4 UUIDs only operate on individual bits.

BTW, if you decide to covert to byte-order-agnostic structure/implementation, please add the unit tests _before_ you convert. You have tested the current implementation quite a bit already so we should use that to test the test cases ;-).

Reply:
I reworked this method to follow the sandard dash-separated [representation](https://datatracker.ietf.org/doc/html/rfc4122#page-3), printing the current layout in host byte order.